### DISCUSSION TITLE
Why are Haskell auto-import code actions unavailable for me?

### DISCUSSION THREAD

**Discussion by emdash-ie** [joaotavora/eglot#1169] [Q&A]
I’m using doom emacs, and recently switched from lsp-mode to eglot to try it out. However, I mostly write Haskell, and some code actions I used to get from lsp-mode don’t appear to be available in eglot.

The main one I’m interested in is auto-imports: with lsp-mode, if I position point on an undefined symbol that GHC can provide an import suggestion for, a code action is available that allows me to choose from one of GHC’s suggested imports, which is then added to the file. With eglot, running `eglot-code-actions` in the same situation tells me there are no code actions here.

Should this work, or are these code actions unsupported by eglot? Looking at other issues suggests this import functionality works when using eglot for go and python, but I think I saw someone asking about this problem with typescript – this makes me think that supporting this action requires work per programming language, though I would’ve assumed the language servers could send the required diff in a language-independent way.

If supporting this in Haskell does require changes to eglot, could anyone point me in the right direction in the code?

In case it’s useful info, I’m using:

- emacs version 28.2, native comp
- doom version 3.0.0-pre
- haskell-language-server version 1.8.0.0, installed with nix
- macOS version 12.6.2

**Comment by 6b6279** [comment]
I cannot reproduce this issue with Emacs 28.2 on Arch Linux, running ghc 9.0.2 and haskell-language-server 1.8.0:
```
module Scratch where
main = List.intercalate ", " ["Lorem", "ipsum", "dolor"]
```
Calling `eglot-code-actions` on `List.intercalate` returns me this list of possible code actions:
![20230220_16h32m06s_grim](https://user-images.githubusercontent.com/109346960/220147399-adf5a189-2a85-43dc-ba0d-770081beb909.png)

It’s possible that eglot won’t recognize possible code actions if your code is malformed or resp. contains invalid syntax elsewhere.

**Comment by emdash-ie** [comment]
Ok, I’m glad to know the code action is supposed to work! I will look into it a little further, try to produce a minimum reproduction. My version of doom is a little old, which might be the issue, and I’m running haskell-server-wrapper through a nix shell based on my project’s cabal file (via cabal2nix), so it could be that. I’ve just tried with your example code in my setup and it’s still not working, so there’s definitely something going on.

**Comment by emdash-ie** [comment]
Have made a repository with a minimum reproduction: https://github.com/undergroundquizscene/eglot-auto-import-code-action-broken

Gonna try updating doom, see if that fixes it.

**Reply by joaotavora** [reply]
Thanks. I will try this.

**Reply by joaotavora** [reply]
Sorry, I just had a look.  I can't try this, it's too much stuff that's way different from my toolchain.  If you can provide a `dockerfile` or a pre-built docker image, maybe I can do this, else I don't think I can sorry.

**Reply by skykanin** [reply]
@joaotavora here is a pre-built docker image with everything setup https://github.com/skykanin/eglot-auto-import-code-action-broken. Instructions are in the `README.md`

**Comment by emdash-ie** [comment]
Nope, updating doom has not worked. Is there any other info I can provide to help reproducing it, or anything else I can try to attempt to fix it? Is it possible to check an eglot version, or is that version now tied to emacs’ own version?

**Reply by 6b6279** [reply]
If you are using Emacs 28.2, eglot isn’t a built-in package yet. But I am guessing that you are using eglot 1.11, the latest version available on ELPA.

One thing I am a bit confused about: I don’t need to invoke the Haskell language server manually if I run eglot when a Haskell file is currently opened. Is it possible that the nix package with the language server is somehow wired differently and eglot is having an hard time interacting with it...? I can’t reproduce your issue, sorry.

**Reply by emdash-ie** [reply]
I believe I have determined that I am using eglot at [this commit](https://github.com/joaotavora/eglot/commit/e501275e06952889056268dabe08ccd0dbaf23e5). I checked the git log in the repository that straight has downloaded for me, and that’s the most recent commit.

I have to invoke the language server manually because eglot doesn’t automatically know how to use nix-shell. It’s entirely possible that running it in the nix shell is the issue – it’s worked this way fine with emacs-lsp, but maybe it doesn’t work with eglot. I’ll try without using nix for the language server and see if that fixes it: if it does then I can try and work out why/how nix breaks it.

Thanks for your help, by the way! I really appreciate it, and I’m sorry I haven’t said until now :)

**Reply by joaotavora** [reply]
Hello Emily, you can try a most recent version of Eglot, which should be installable in Emacs 28.2  if you add "GNU-devel ELPA" to your Emacs package repositories:

```
(add-to-list 'package-archives '("gnu-devel" . "https://elpa.gnu.org/devel/"))
```

After this, `M-x package-refresh-contents` and `M-x package-list-packages` should bing up a menu.

Alternatively, you can wait a bit for Eglot 1.12 to be released to GNU ELPA.

**Comment by emdash-ie** [comment]
I haven’t worked out my problem with this yet, but to update this discussion for the record (in case it might be useful to someone else), I’ve determined that I’m using roughly version 1.9 of eglot. I think for now I’ll return to using lsp-mode until emacs 29 is out and I can get newer versions of eglot more straightforwardly. Since I’m using doom emacs, I’m not entirely sure if adding GNU-devel ELPA will work, and I don’t want to mess with it too much.

**Reply by skykanin** [reply]
I'm also experiencing this same issue when using doom emacs, providing `haskell-language-server-wrapper` through a nix shell and using `envrc` together with direnv to let emacs get access to the HLS binary. You should be able to specify the eglot commit you want in the `packages.el` file like so `(package! eglot :pin "8b5532dd32b25276c1857508030b207f765ef9b6")`.

For the record I'm pinned to `8b5532dd32b25276c1857508030b207f765ef9b6` which is eglot 1.15 and still experiencing this issue.
